File: frontend/admin/packages/modules/system/src/api/system/admin/v1/base_rate_limit_rule.ts
import service from "@liujitcn/kratos-admin-core/request";
import type {
  BaseRateLimitRuleForm,
  BaseRateLimitRuleService,
  CreateBaseRateLimitRuleRequest,
  DeleteBaseRateLimitRuleRequest,
  GetBaseRateLimitRuleRequest,
  OptionBaseRateLimitRuleResponse,
  OptionBaseRateLimitRuleRequest,
  PageBaseRateLimitRuleRequest,
  PageBaseRateLimitRuleResponse,
  SetBaseRateLimitRuleStatusRequest,
  UpdateBaseRateLimitRuleRequest
} from "@liujitcn/kratos-admin-system/rpc/system/admin/v1/base_rate_limit_rule";
import type { Empty } from "@liujitcn/kratos-admin-system/rpc/google/protobuf/empty";

const BASE_RATE_LIMIT_RULE_URL = "/v1/admin/base/rate-limit-rule";

/** 管理端限流规则模板服务。 */
export class BaseRateLimitRuleServiceImpl implements BaseRateLimitRuleService {
  /** 查询限流规则选项。 */
  OptionBaseRateLimitRule(request: OptionBaseRateLimitRuleRequest = { keyword: "" }): Promise<OptionBaseRateLimitRuleResponse> {
    return service<OptionBaseRateLimitRuleRequest, OptionBaseRateLimitRuleResponse>({ url: `${BASE_RATE_LIMIT_RULE_URL}/option`, method: "get", params: request });
  }

  /** 查询限流规则分页列表。 */
  PageBaseRateLimitRule(request: PageBaseRateLimitRuleRequest): Promise<PageBaseRateLimitRuleResponse> {
    return service<PageBaseRateLimitRuleRequest, PageBaseRateLimitRuleResponse>({ url: BASE_RATE_LIMIT_RULE_URL, method: "get", params: request });
  }

  /** 查询限流规则详情。 */
  GetBaseRateLimitRule(request: GetBaseRateLimitRuleRequest): Promise<BaseRateLimitRuleForm> {
    return service<GetBaseRateLimitRuleRequest, BaseRateLimitRuleForm>({ url: `${BASE_RATE_LIMIT_RULE_URL}/${request.id}`, method: "get" });
  }

  /** 创建限流规则。 */
  CreateBaseRateLimitRule(request: CreateBaseRateLimitRuleRequest): Promise<Empty> {
    return service<BaseRateLimitRuleForm | undefined, Empty>({ url: BASE_RATE_LIMIT_RULE_URL, method: "post", data: request.base_rate_limit_rule });
  }

  /** 更新限流规则。 */
  UpdateBaseRateLimitRule(request: UpdateBaseRateLimitRuleRequest): Promise<Empty> {
    return service<BaseRateLimitRuleForm | undefined, Empty>({ url: `${BASE_RATE_LIMIT_RULE_URL}/${request.base_rate_limit_rule?.id ?? ""}`, method: "put", data: request.base_rate_limit_rule });
  }

  /** 删除限流规则。 */
  DeleteBaseRateLimitRule(request: DeleteBaseRateLimitRuleRequest): Promise<Empty> {
    return service<DeleteBaseRateLimitRuleRequest, Empty>({ url: `${BASE_RATE_LIMIT_RULE_URL}/${request.id}`, method: "delete" });
  }

  /** 设置限流规则状态。 */
  SetBaseRateLimitRuleStatus(request: SetBaseRateLimitRuleStatusRequest): Promise<Empty> {
    return service<SetBaseRateLimitRuleStatusRequest, Empty>({ url: `${BASE_RATE_LIMIT_RULE_URL}/${request.id}/status`, method: "put", data: request });
  }
}

/** 限流规则模板服务实例。 */
export const defBaseRateLimitRuleService = new BaseRateLimitRuleServiceImpl();

File: backend/internal/service/system/admin/v1/base_rate_limit_rule.go
package admin

import (
	"context"
	"fmt"

	"github.com/go-kratos/kratos/v3/log"
	adminv1 "github.com/liujitcn/kratos-admin/backend/api/gen/go/system/admin/v1"
	biz "github.com/liujitcn/kratos-admin/backend/internal/biz/system/admin"
	"github.com/liujitcn/kratos-core/errorsx"
	"google.golang.org/protobuf/types/known/emptypb"
)

// BaseRateLimitRuleService 提供限流规则模板管理接口。
type BaseRateLimitRuleService struct {
	adminv1.UnimplementedBaseRateLimitRuleServiceServer
	baseRateLimitRuleCase *biz.BaseRateLimitRuleCase
}

// NewBaseRateLimitRuleService 创建限流规则模板服务。
func NewBaseRateLimitRuleService(baseRateLimitRuleCase *biz.BaseRateLimitRuleCase) *BaseRateLimitRuleService {
	return &BaseRateLimitRuleService{baseRateLimitRuleCase: baseRateLimitRuleCase}
}

// OptionBaseRateLimitRule 查询限流规则选项。
func (s *BaseRateLimitRuleService) OptionBaseRateLimitRule(ctx context.Context, req *adminv1.OptionBaseRateLimitRuleRequest) (*adminv1.OptionBaseRateLimitRuleResponse, error) {
	result, err := s.baseRateLimitRuleCase.OptionBaseRateLimitRule(ctx, req)
	if err != nil {
		log.Error(fmt.Sprintf("OptionBaseRateLimitRule %v", err))
		return nil, errorsx.WrapInternal(err, "查询限流规则选项失败")
	}
	return result, nil
}

// PageBaseRateLimitRule 分页查询限流规则模板。
func (s *BaseRateLimitRuleService) PageBaseRateLimitRule(ctx context.Context, req *adminv1.PageBaseRateLimitRuleRequest) (*adminv1.PageBaseRateLimitRuleResponse, error) {
	result, err := s.baseRateLimitRuleCase.PageBaseRateLimitRule(ctx, req)
	if err != nil {
		log.Error(fmt.Sprintf("PageBaseRateLimitRule %v", err))
		return nil, errorsx.WrapInternal(err, "分页查询限流规则失败")
	}
	return result, nil
}

// GetBaseRateLimitRule 查询限流规则模板详情。
func (s *BaseRateLimitRuleService) GetBaseRateLimitRule(ctx context.Context, req *adminv1.GetBaseRateLimitRuleRequest) (*adminv1.BaseRateLimitRuleForm, error) {
	result, err := s.baseRateLimitRuleCase.GetBaseRateLimitRule(ctx, req.GetId())
	if err != nil {
		log.Error(fmt.Sprintf("GetBaseRateLimitRule %v", err))
		return nil, errorsx.WrapInternal(err, "查询限流规则详情失败")
	}
	return result, nil
}

// CreateBaseRateLimitRule 创建限流规则模板。
func (s *BaseRateLimitRuleService) CreateBaseRateLimitRule(ctx context.Context, req *adminv1.CreateBaseRateLimitRuleRequest) (*emptypb.Empty, error) {
	err := s.baseRateLimitRuleCase.CreateBaseRateLimitRule(ctx, req.GetBaseRateLimitRule())
	if err != nil {
		log.Error(fmt.Sprintf("CreateBaseRateLimitRule %v", err))
		return nil, errorsx.WrapInternal(err, "创建限流规则失败")
	}
	return new(emptypb.Empty), nil
}

// UpdateBaseRateLimitRule 更新限流规则模板。
func (s *BaseRateLimitRuleService) UpdateBaseRateLimitRule(ctx context.Context, req *adminv1.UpdateBaseRateLimitRuleRequest) (*emptypb.Empty, error) {
	err := s.baseRateLimitRuleCase.UpdateBaseRateLimitRule(ctx, req.GetBaseRateLimitRule())
	if err != nil {
		log.Error(fmt.Sprintf("UpdateBaseRateLimitRule %v", err))
		return nil, errorsx.WrapInternal(err, "更新限流规则失败")
	}
	return new(emptypb.Empty), nil
}

// DeleteBaseRateLimitRule 删除限流规则模板。
func (s *BaseRateLimitRuleService) DeleteBaseRateLimitRule(ctx context.Context, req *adminv1.DeleteBaseRateLimitRuleRequest) (*emptypb.Empty, error) {
	err := s.baseRateLimitRuleCase.DeleteBaseRateLimitRule(ctx, req.GetId())
	if err != nil {
		log.Error(fmt.Sprintf("DeleteBaseRateLimitRule %v", err))
		return nil, errorsx.WrapInternal(err, "删除限流规则失败")
	}
	return new(emptypb.Empty), nil
}

// SetBaseRateLimitRuleStatus 设置限流规则模板状态。
func (s *BaseRateLimitRuleService) SetBaseRateLimitRuleStatus(ctx context.Context, req *adminv1.SetBaseRateLimitRuleStatusRequest) (*emptypb.Empty, error) {
	err := s.baseRateLimitRuleCase.SetBaseRateLimitRuleStatus(ctx, req)
	if err != nil {
		log.Error(fmt.Sprintf("SetBaseRateLimitRuleStatus %v", err))
		return nil, errorsx.WrapInternal(err, "设置限流规则状态失败")
	}
	return new(emptypb.Empty), nil
}
